### PR TITLE
gui: Drop qt4 support

### PR DESCRIPTION
**Pull request by laanwj**
Implements #8263.

Qt4.x has been EOL since 2015, and at least Gentoo has, or is going to drop support for it. I wouldn't be surprised if other Linux distributions follow.

This removes Qt4 detection from the build system, as well as removes all Qt4 fallbacks from the code. Turns out there's more than I expected: this is going to make maintenance of the GUI code, as well as adding new features significantly easier.

(I know there's still some references left to qt4 in RPM and Debian build script, but I don't have the knowledge how to fix them)

### CODE REVIEW

**Comment by sipa** [comment]
Typo in the title.

**Comment by fanquake** [comment]
What would this make our minimum supported version of Qt? 5.4.x IIRC?

**Comment by laanwj** [comment]
> What would this make our minimum supported version of Qt? 5.4.x IIRC?

That's a good question. I'm not sure that's something to be decided here, I don't change anything with regard to Qt5 versions supported. Though there are a few `QT_VERSION` checks for Qt 5.2 that would not be relevant in that case..

File: build-aux/m4/bitcoin_qt.m4 [hunk 1]
```diff
@@ -113,63 +113,48 @@ AC_DEFUN([BITCOIN_QT_CONFIGURE],[
   TEMP_CXXFLAGS=$CXXFLAGS
```
Comment (fanquake) [inline build-aux/m4/bitcoin_qt.m4]: Qt4 mentions in the comment here ^, and another starting at line 288.

**Comment by fanquake** [comment]
> That's a good question. I'm not sure that's something to be decided here, I don't change anything with regard to Qt5 versions supported. Though there are a few QT_VERSION checks for Qt 5.2 that would not be relevant in that case..

Fair enough.

Still some qt4 related code here:
https://github.com/bitcoin/bitcoin/blob/master/src/qt/test/wallettests.cpp#L90

There are some qt4 related docs that could be dropped. i.e:
https://github.com/bitcoin/bitcoin/blame/master/doc/build-osx.md#L27
https://github.com/bitcoin/bitcoin/blob/master/doc/build-unix.md#dependencies-for-the-gui

**Comment by laanwj** [comment]
Thanks @fanquake, updated for those.

**Comment by promag** [comment]
Concept ACK, will review.

**Comment by MarcoFalke** [comment]
ACK. A follow up pull request could clarify which versions of qt5 are supported.

**Comment by MarcoFalke** [comment]
Nit: The section for fedora still says " To build with Qt 5 (recommended) you need the following:"

**Comment by theuni** [comment]
Concept ACK.

File: .travis.yml [hunk 2]
```diff
@@ -35,8 +35,8 @@ env:
...
-    - HOST=x86_64-unknown-linux-gnu PACKAGES="python3-zmq qt4-dev-tools libssl1.0-dev libevent-dev bsdmainutils libboost-system-dev libboost-filesystem-dev libboost-chrono-dev libboost-program-options-dev libboost-test-dev libboost-thread-dev libdb5.3++-dev libminiupnpc-dev libzmq3-dev libprotobuf-dev protobuf-compiler libqrencode-dev xvfb libqt4-dev" NO_DEPENDS=1 NEED_XVFB=1 RUN_TESTS=true GOAL="install" BITCOIN_CONFIG="--enable-zmq --with-incompatible-bdb --enable-glibc-back-compat --enable-reduce-exports --with-gui=qt4 CPPFLAGS=-DDEBUG_LOCKORDER" DISPLAY=:99.0
+# x86_64 Linux (Qt5 & system libs)
+    - HOST=x86_64-unknown-linux-gnu PACKAGES="python3-zmq qtbase5-dev qttools5-dev-tools libssl1.0-dev libevent-dev bsdmainutils libboost-system-dev libboost-filesystem-dev libboost-chrono-dev libboost-program-options-dev libboost-test-dev libboost-thread-dev libdb5.3++-dev libminiupnpc-dev libzmq3-dev libprotobuf-dev protobuf-compiler libqrencode-dev xvfb libqt4-dev" NO_DEPENDS=1 NEED_XVFB=1 RUN_TESTS=true GOAL="install" BITCOIN_CONFIG="--enable-zmq --with-incompatible-bdb --enable-glibc-back-compat --enable-reduce-exports --with-gui=qt5 CPPFLAGS=-DDEBUG_LOCKORDER" DISPLAY=:99.0
```
Comment (theuni) [inline .travis.yml]: Missed libqt4-dev.

Also, I believe this means that we can get rid of xvfb, since qt5's 'minimal' plugin should work.

**Comment by MeshCollider** [comment]
Concept ACK

File: src/qt/guiutil.h [hunk 3]
```diff
@@ -233,7 +233,7 @@ namespace GUIUtil
         void mouseReleaseEvent(QMouseEvent *event);
     };
 
-#if defined(Q_OS_MAC) && QT_VERSION >= 0x050000
+#if defined(Q_OS_MAC)
```
Comment (fanquake) [inline src/qt/guiutil.h]: Note: Still need to test, but we might be able to remove this Qt workaround entirely.

**Comment by jonasschnelli** [comment]
Concept ACK
Gitian Build: https://bitcoin.jonasschnelli.ch/build/656

**Comment by promag** [comment]
Remove `Qt 4 is also supported` from `bitcoin/src/qt/README.md`.

**Comment by promag** [comment]
All `QT_VERSION` usages LGTM.

**Comment by laanwj** [comment]
Ok, I think all comments by @promag @MarcoFalke @theuni have been addressed.
@fanquake That'd be good. But maybe in a separate PR, I think this update should be dumbly removing Qt4 support but otherwise keep the logic for Qt5 the same.

File: .travis.yml [hunk 4]
```diff
@@ -35,8 +35,8 @@ env:
...
-# Qt4 & system libs
-    - HOST=x86_64-unknown-linux-gnu PACKAGES="python3-zmq qt4-dev-tools libssl1.0-dev libevent-dev bsdmainutils libboost-system-dev libboost-filesystem-dev libboost-chrono-dev libboost-program-options-dev libboost-test-dev libboost-thread-dev libdb5.3++-dev libminiupnpc-dev libzmq3-dev libprotobuf-dev protobuf-compiler libqrencode-dev xvfb libqt4-dev" NO_DEPENDS=1 NEED_XVFB=1 RUN_TESTS=true GOAL="install" BITCOIN_CONFIG="--enable-zmq --with-incompatible-bdb --enable-glibc-back-compat --enable-reduce-exports --with-gui=qt4 CPPFLAGS=-DDEBUG_LOCKORDER" DISPLAY=:99.0
+# x86_64 Linux (Qt5 & system libs)
+    - HOST=x86_64-unknown-linux-gnu PACKAGES="python3-zmq qtbase5-dev qttools5-dev-tools libssl1.0-dev libevent-dev bsdmainutils libboost-system-dev libboost-filesystem-dev libboost-chrono-dev libboost-program-options-dev libboost-test-dev libboost-thread-dev libdb5.3++-dev libminiupnpc-dev libzmq3-dev libprotobuf-dev protobuf-compiler libqrencode-dev" NO_DEPENDS=1 NEED_XVFB=1 RUN_TESTS=true GOAL="install" BITCOIN_CONFIG="--enable-zmq --with-incompatible-bdb --enable-glibc-back-compat --enable-reduce-exports --with-gui=qt5 CPPFLAGS=-DDEBUG_LOCKORDER" DISPLAY=:99.0
```
Comment (MarcoFalke) [inline .travis.yml]: I think cory mentioned you could get rid of xvfb?

Comment (laanwj) [reply .travis.yml]: Didn't I? oh, there's an environment variable too...

Comment (theuni) [reply .travis.yml]: $DISPLAY was also related to xvfb and can be removed.

File: .travis.yml [hunk 5]
```diff
@@ -35,8 +35,8 @@ env:
...
-# Qt4 & system libs
-    - HOST=x86_64-unknown-linux-gnu PACKAGES="python3-zmq qt4-dev-tools libssl1.0-dev libevent-dev bsdmainutils libboost-system-dev libboost-filesystem-dev libboost-chrono-dev libboost-program-options-dev libboost-test-dev libboost-thread-dev libdb5.3++-dev libminiupnpc-dev libzmq3-dev libprotobuf-dev protobuf-compiler libqrencode-dev xvfb libqt4-dev" NO_DEPENDS=1 NEED_XVFB=1 RUN_TESTS=true GOAL="install" BITCOIN_CONFIG="--enable-zmq --with-incompatible-bdb --enable-glibc-back-compat --enable-reduce-exports --with-gui=qt4 CPPFLAGS=-DDEBUG_LOCKORDER" DISPLAY=:99.0
+# x86_64 Linux (Qt5 & system libs)
+    - HOST=x86_64-unknown-linux-gnu PACKAGES="python3-zmq qtbase5-dev qttools5-dev-tools libssl1.0-dev libevent-dev bsdmainutils libboost-system-dev libboost-filesystem-dev libboost-chrono-dev libboost-program-options-dev libboost-test-dev libboost-thread-dev libdb5.3++-dev libminiupnpc-dev libzmq3-dev libprotobuf-dev protobuf-compiler libqrencode-dev" NO_DEPENDS=1 RUN_TESTS=true GOAL="install" BITCOIN_CONFIG="--enable-zmq --with-incompatible-bdb --enable-glibc-back-compat --enable-reduce-exports --with-gui=qt5 CPPFLAGS=-DDEBUG_LOCKORDER" DISPLAY=:99.0
```
Comment (MarcoFalke) [inline .travis.yml]: Can probably also remove two instances of `DISPLAY`?

**Comment by practicalswift** [comment]
Concept ACK

**Comment by MarcoFalke** [comment]
@laanwj You wouldn't need to export `DISPLAY` to the docker as well? Could remove that as well?

**Comment by fanquake** [comment]
Should be able to drop this check from [paymentrequestplus.cpp](https://github.com/bitcoin/bitcoin/blob/master/src/qt/paymentrequestplus.cpp#L100):
```
#if QT_VERSION >= 0x050000
        if (qCert.isBlacklisted()) {
            ...snip...
            return false;
        }
#endif
```

**Comment by laanwj** [comment]
@MarcoFalke @fanquake done
this is becoming a palimpsest of commits, if there's no more things I missed I'm going to squash them into more sensible subdivision

**Comment by TheBlueMatt** [comment]
I presume, barring some major rewrite, this implies the PPA should simply not ship with a GUI except on 18.04? (And replace bitcoin-qt with a dummy package instead of letting it get stale?)

**Comment by MarcoFalke** [comment]
@TheBlueMatt I doubt anyone running ubuntu 14.04 is running the gui from the ppa. If there is one, they might as well download the release build.

**Comment by MarcoFalke** [comment]
Ah, according to #8043, using the release builds wouldn't help and xenial is affected as well. Though, since the only problem is the tray icon, we might as well just ship a package with a slight ui glitch instead of no package at all.

**Comment by fanquake** [comment]
@wumpus That'd be great.

**Comment by laanwj** [comment]
> @TheBlueMatt I doubt anyone running ubuntu 14.04 is running the gui from the ppa. If there is one, they might as well download the release build.

Or a three-line patch to disable tray icon support in the PPA, for those versions. It seems ridiculous to me to remove the entire GUI package just because of a small glitch.

**Comment by wumpus** [comment]
Wrong @wumpus.

**Comment by laanwj** [comment]
Squashed into sensible commits:
56c25b36b17b73b4ad36d608c428aaa052130fab → af6ac3b677454644364fd24d0df0c02ac9b8c8db 
No other changes.

**Comment by MarcoFalke** [comment]
Concept re-ACK af6ac3b677. (No changes since feedback addressed, but I didn't review the build changes)

**Pull request merged**